File: src/mapscript/perl/examples/RFC24.pl
#!/usr/bin/perl
use strict;
use warnings;
use mapscript;
use Data::Dumper;

my $file=$ARGV[0];

# utility
sub assertNotNull {
	my ($o, $test) = @_;
	if ($o) {
		print $test . "  PASSED\n";
	} else {
		print $test . "  FAILED\n";
	}
}


# layerObj
sub testGetLayerObj {
	my $map = new mapscript::mapObj($file);
	my $layer = $map->getLayer(1);
	$map = undef;
	assertNotNull( $layer->{map} , "testGetLayerObj");
	#$layer->{map}->draw()->save("/tmp/map.png");
}

sub testGetLayerObjByName {
	my $map = new mapscript::mapObj($file);
	my $layer = $map->getLayerByName("POLYGON");
	$map = undef;
	assertNotNull( $layer->{map} , "testGetLayerObjByName");
}

sub testLayerObj {
	my $map = new mapscript::mapObj($file);
	my $layer = new mapscript::layerObj($map);
	$map = undef;
	assertNotNull( $layer->{map} , "testLayerObj");
}

sub testInsertLayerObj {
	my $map = new mapscript::mapObj($file);
	my $layer = new mapscript::layerObj(undef);
	my $position = $map->insertLayer($layer);
	$map = undef;
	assertNotNull( $position == 7 , "testInsertLayerObj position");
	assertNotNull( $layer->{map} , "testInsertLayerObj notnull");
}

# classObj
sub testGetClassObj {
	#dumpHash(mapscript::getPARENT_PTRS());
	my $map = new mapscript::mapObj($file);
	my $layer = $map->getLayer(1);
	my $clazz = $layer->getClass(0);
	#dumpHash(mapscript::getPARENT_PTRS());
	#$clazz->{layer}->{map}->draw()->save("/tmp/map1.png");
	#print "parent layer=".$clazz->{layer}."\n";
	$map = undef; $layer=undef;
	#print "parent layer=".$clazz->{layer}."\n";
	assertNotNull( $clazz->{layer} , "testGetClassObj");
	#$clazz->{layer}->{map}->draw()->save("/tmp/map2.png");
	#dumpHash(mapscript::getPARENT_PTRS());
}

sub testClassObj {
	my $map = new mapscript::mapObj($file);
	my $layer = $map->getLayer(1);
	my $clazz = new mapscript::classObj($layer);
	$map = undef; $layer=undef;
	assertNotNull( $clazz->{layer} , "testClassObj");
}

sub testInsertClassObj {
	my $map = new mapscript::mapObj($file);
	my $layer = $map->getLayer(1);
	my $clazz = new mapscript::classObj(undef);
	my $position = $layer->insertClass($clazz);
	$map = undef; $layer=undef;
	assertNotNull( $position == 2 , "testInsertClassObj position");
	assertNotNull( $clazz->{layer} , "testInsertClassObj notnull");
}

if ( ! $file ) {
	print "Usage: RFC24.pl file.map\n";
	exit 1;
}

testGetLayerObj;
testGetLayerObjByName;
testLayerObj;
testInsertLayerObj;
testGetClassObj;
testClassObj;
testInsertClassObj;

my $hashmap = mapscript::getPARENT_PTRS();
assertNotNull( keys( %$hashmap )==0 , "checking that hashmap of parent ptrs is empty");
print "No of keys:".keys( %$hashmap )."\n";
#dumpHash($hashmap);
print Dumper( $hashmap );

File: src/mapscript/java/examples/RFC24.java
import edu.umn.gis.mapscript.*;

/**
 * <p>Title: Mapscript RFC24 tests.</p>
 * <p>Description: Tests for RFC24 implementation. (http://mapserver.gis.umn.edu/development/rfc/ms-rfc-24/)</p>
 * @author Umberto Nicoletti (umberto.nicoletti@gmail.com)
 */

public class RFC24 {
	String mapfile;

	public static void main(String[] args) {
		new RFC24(args[0]).run();
	}
	
	public RFC24(String mapfile) {
		this.mapfile=mapfile;
	}
	
	public void run() {
		System.out.println("Running "+getClass().getName());
		testLayerObj();
		testClassObj();
		testInsertLayerObj();
		testInsertClassObj();
		testGetLayerObj();
		testGetLayerObjByName();
		testGetClassObj();
		testGetOutputformatObj();
		testInsertOutputformatObj();

		System.out.println("Finished "+getClass().getName());
	}
	
	public void testLayerObj() {
		mapObj map=new mapObj(mapfile);
		layerObj newLayer=new layerObj(map);
		
		map=null;
		gc();
		assertNotNull(newLayer.getMap(), "testLayerObj");
	}
	
	public void testInsertLayerObj() {
		mapObj map=new mapObj(mapfile);
		layerObj newLayer=new layerObj(null);
		map.insertLayer(newLayer,-1);
		
		map=null;
		gc();
		assertNotNull(newLayer.getMap(), "testInsertLayerObj");
	}
	
	public void testGetLayerObj() {
		mapObj map=new mapObj(mapfile);
		layerObj newLayer=map.getLayer(1);
		
		map=null;
		gc();
		assertNotNull(newLayer.getMap(), "testGetLayerObj");
	}

	public void testGetLayerObjByName() {
		mapObj map=new mapObj(mapfile);
		layerObj newLayer=map.getLayerByName("POLYGON");
		
		map=null;
		gc();
		assertNotNull(newLayer.getMap(), "testGetLayerObjByName");
	}

	public void testClassObj() {
		mapObj map=new mapObj(mapfile);
		layerObj layer=map.getLayer(1);
		classObj newClass=new classObj(layer);
		
		map=null; layer=null;
		gc();
		assertNotNull(newClass.getLayer(), "testClassObj");
	}

	public void testInsertClassObj() {
		mapObj map=new mapObj(mapfile);
		layerObj layer=map.getLayer(1);
		classObj newClass=new classObj(null);
		layer.insertClass(newClass,-1);
		
		map=null; layer=null;
		gc();
		assertNotNull(newClass.getLayer(), "testInsertClassObj");
	}

	public void testGetClassObj() {
		mapObj map=new mapObj(mapfile);
		layerObj layer=map.getLayer(1);
		classObj newClass=layer.getClass(0);
		
		map=null; layer=null;
		gc();
		assertNotNull(newClass.getLayer(), "testGetClassObj");
	}

        public void testGetOutputformatObj() {
                mapObj map=new mapObj(mapfile);
                outputFormatObj format=null;
                for (int i=0; i<map.getNumoutputformats(); i++) {
                    format = map.getOutputFormat(i);
                    //System.out.println("["+i+"] Format name: "+format.getName());
                } 
                map=null;
                gc();
                assertNotNull(format.getDriver(), "testGetOutputformatObj");
        }

        public void testInsertOutputformatObj() {
                mapObj map=new mapObj(mapfile);
                outputFormatObj format=null;
                for (int i=0; i<map.getNumoutputformats(); i++) {
                    format = map.getOutputFormat(i);
		    // uncomment to see output
                    //System.out.println("["+i+"] Format name: "+format.getName() + " driver="+format.getDriver());
                } 
                outputFormatObj newFormat=new outputFormatObj("AGG/PNG","test/rfc24");
		map.appendOutputFormat(newFormat);
                for (int i=0; i<map.getNumoutputformats(); i++) {
                    format = map.getOutputFormat(i);
		    // uncomment to see output
                    //System.out.println("["+i+"] Format name: "+format.getName());
                } 
		map.draw();
                map=null; format=null;
                gc();
                assertNotNull(newFormat.getDriver(), "testInsertOutputformatObj");
        }

	public void gc() {
		for (int i=0; i<10; i++)
			System.gc();
	}
	
	public void assertNotNull(Object object, String test) {
		if ( object != null )
			System.out.println("\t- "+test+" PASSED");
		else
			System.out.println("\t- "+test+" FAILED");
	}
}
